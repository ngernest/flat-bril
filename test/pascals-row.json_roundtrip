{
  "functions": [
    {
      "args": [],
      "instrs": [
        {
          "dest": "v0",
          "op": "const",
          "type": "int",
          "value": 5
        },
        {
          "args": [
            "v0"
          ],
          "dest": "x",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "x"
          ],
          "dest": "v1",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "v1"
          ],
          "funcs": [
            "generateNthRow"
          ],
          "labels": [],
          "op": "call"
        },
        {
          "dest": "v2",
          "op": "const",
          "type": "int",
          "value": 0
        }
      ],
      "name": "main"
    },
    {
      "args": [
        {
          "name": "x",
          "type": "int"
        }
      ],
      "instrs": [
        {
          "dest": "v0",
          "op": "const",
          "type": "int",
          "value": 1
        },
        {
          "args": [
            "v0"
          ],
          "dest": "prev",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "prev"
          ],
          "dest": "v1",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "v1"
          ],
          "funcs": [],
          "labels": [],
          "op": "print"
        },
        {
          "dest": "v2",
          "op": "const",
          "type": "int",
          "value": 0
        },
        {
          "dest": "v4",
          "op": "const",
          "type": "int",
          "value": 0
        },
        {
          "args": [
            "v4"
          ],
          "dest": "i",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "label": "for.cond.3"
        },
        {
          "args": [
            "i"
          ],
          "dest": "v5",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "x"
          ],
          "dest": "v6",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "v5",
            "v6"
          ],
          "dest": "v7",
          "funcs": [],
          "labels": [],
          "op": "lt",
          "type": "bool"
        },
        {
          "args": [
            "v7"
          ],
          "funcs": [],
          "labels": [
            "for.body.3",
            "for.end.3"
          ],
          "op": "br"
        },
        {
          "label": "for.body.3"
        },
        {
          "args": [
            "x"
          ],
          "dest": "v8",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "i"
          ],
          "dest": "v9",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "v8",
            "v9"
          ],
          "dest": "v10",
          "funcs": [],
          "labels": [],
          "op": "sub",
          "type": "int"
        },
        {
          "args": [
            "v10"
          ],
          "dest": "temp",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "temp"
          ],
          "dest": "v11",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "dest": "v12",
          "op": "const",
          "type": "int",
          "value": 1
        },
        {
          "args": [
            "v11",
            "v12"
          ],
          "dest": "v13",
          "funcs": [],
          "labels": [],
          "op": "add",
          "type": "int"
        },
        {
          "args": [
            "v13"
          ],
          "dest": "temp",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "prev"
          ],
          "dest": "v14",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "temp"
          ],
          "dest": "v15",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "v14",
            "v15"
          ],
          "dest": "v16",
          "funcs": [],
          "labels": [],
          "op": "mul",
          "type": "int"
        },
        {
          "args": [
            "v16"
          ],
          "dest": "curr",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "curr"
          ],
          "dest": "v17",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "v17"
          ],
          "funcs": [],
          "labels": [],
          "op": "print"
        },
        {
          "dest": "v18",
          "op": "const",
          "type": "int",
          "value": 0
        },
        {
          "args": [
            "curr"
          ],
          "dest": "v19",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "v19"
          ],
          "dest": "prev",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "i"
          ],
          "dest": "v20",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "dest": "v21",
          "op": "const",
          "type": "int",
          "value": 1
        },
        {
          "args": [
            "v20",
            "v21"
          ],
          "dest": "v22",
          "funcs": [],
          "labels": [],
          "op": "add",
          "type": "int"
        },
        {
          "args": [
            "v22"
          ],
          "dest": "i",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [],
          "funcs": [],
          "labels": [
            "for.cond.3"
          ],
          "op": "jmp"
        },
        {
          "label": "for.end.3"
        }
      ],
      "name": "generateNthRow"
    }
  ]
}
