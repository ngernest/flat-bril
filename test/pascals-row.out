@main
	op: const		dest: v0		type: int		value: 5	
	op: id   		dest: x 		type: int		args: ["v0"]	
	op: id   		dest: v1		type: int		args: ["x"]	
	op: call 		args: ["v1"]		funcs: ["generateNthRow"]
	op: const		dest: v2		type: int		value: 0	
@generateNthRow
	op: const		dest: v0		type: int		value: 1	
	op: id   		dest: prev		type: int		args: ["v0"]	
	op: id   		dest: v1		type: int		args: ["prev"]	
	op: print		args: ["v1"]	
	op: const		dest: v2		type: int		value: 0	
	op: const		dest: v4		type: int		value: 0	
	op: id   		dest: i 		type: int		args: ["v4"]	
.for.cond.3
	op: id   		dest: v5		type: int		args: ["i"]	
	op: id   		dest: v6		type: int		args: ["x"]	
	op: lt   		dest: v7		type: bool		args: ["v5", "v6"]	
	op: br   		args: ["v7"]		labels: ["for.body.3", "for.end.3"]
.for.body.3
	op: id   		dest: v8		type: int		args: ["x"]	
	op: id   		dest: v9		type: int		args: ["i"]	
	op: sub  		dest: v10		type: int		args: ["v8", "v9"]	
	op: id   		dest: temp		type: int		args: ["v10"]	
	op: id   		dest: v11		type: int		args: ["temp"]	
	op: const		dest: v12		type: int		value: 1	
	op: add  		dest: v13		type: int		args: ["v11", "v12"]	
	op: id   		dest: temp		type: int		args: ["v13"]	
	op: id   		dest: v14		type: int		args: ["prev"]	
	op: id   		dest: v15		type: int		args: ["temp"]	
	op: mul  		dest: v16		type: int		args: ["v14", "v15"]	
	op: id   		dest: curr		type: int		args: ["v16"]	
	op: id   		dest: v17		type: int		args: ["curr"]	
	op: print		args: ["v17"]	
	op: const		dest: v18		type: int		value: 0	
	op: id   		dest: v19		type: int		args: ["curr"]	
	op: id   		dest: prev		type: int		args: ["v19"]	
	op: id   		dest: v20		type: int		args: ["i"]	
	op: const		dest: v21		type: int		value: 1	
	op: add  		dest: v22		type: int		args: ["v20", "v21"]	
	op: id   		dest: i 		type: int		args: ["v22"]	
	op: jmp  		labels: ["for.cond.3"]
.for.end.3
